### DISCUSSION TITLE
Any example about TableRelativePermeability?

### DISCUSSION THREAD

**Discussion by wdmhouston** [GEOS-DEV/GEOS#1480] [Q&A] [answered]
Is there any example of how to use TableRelativePermeability? I notice that tablefunction is required, but couldn't find any example online.

**Comment by klevzoff** [comment]
Hi @wdmhouston ,

At the moment, we only have one example in the repository using 3-phase table-based relperms:
https://github.com/GEOSX/GEOSX/blob/develop/src/coreComponents/physicsSolvers/fluidFlow/wells/integratedTests/compositionalMultiphaseWell/dead_oil_wells_2d.xml

The usage is as follows:
```xml
  <Constitutive>

    <TableRelativePermeability
      name="relperm"
      phaseNames="{ oil, gas, water }"
      waterOilRelPermTableNames="{ waterRelPermTable, oilRelPermTableForOW }"
      gasOilRelPermTableNames="{ gasRelPermTable, oilRelPermTableForOG }"/>

  </Constitutive>

  <Functions>

    <TableFunction
      name="waterRelPermTable"
      coordinateFiles="{ satw.txt }"
      voxelFile="krw.txt"
      interpolation="linear"/>

    <TableFunction
      name="oilRelPermTableForOW"
      coordinateFiles="{ sato.txt }"
      voxelFile="krow.txt"
      interpolation="linear"/>

    <TableFunction
      name="gasRelPermTable"
      coordinateFiles="{ satg.txt }"
      voxelFile="krg.txt"
      interpolation="linear"/>

    <TableFunction
      name="oilRelPermTableForOG"
      coordinateFiles="{ sato.txt }"
      voxelFile="krog.txt"
      interpolation="linear"/>
  
  </Functions>
```

where `satw.txt`, `krw.txt`, etc. are plain text files residing in the same directory as the input file, containing one value (saturation or relperm, respectively) per line.

Instead of separate text files, you can also specify the table inline using attributes `coordinates` (for saturation) and `values` (for relperm):
```xml
<TableFunction
      name="waterRelPermTable"
      coordinates="{ 0.05, 0.5, 1 }"
      values="{ 0, 0.4, 0.9 }"
      interpolation="linear"/>
```

**Comment by klevzoff** [comment]
I've created an issue to document the inputs for this type of models: https://github.com/GEOSX/GEOSX/issues/1481

**Reply by wdmhouston** [reply]
Thanks! will try!

**Reply by wdmhouston** [reply]
for CO2 case, I have co2 and water only.  Should I use waterOilRelPermTableNames or gasOilRelPermTableNames?

<TableRelativePermeability
			name="relperm"
			phaseNames="{ co2, water }"
             ?          waterOilRelPermTableNames="{ waterRelPermTable, oilRelPermTableForOW }" 
             ?         gasOilRelPermTableNames="{ gasRelPermTable, oilRelPermTableForOG }"/>
...

**Reply by klevzoff** [reply]
Unfortunately, this model does not work without a (reference) oil phase at the moment. So it can support 2-phase gas-oil and water-oil systems, but not water-gas. I've created another issue, https://github.com/GEOSX/GEOSX/issues/1483 , to extend the model to this case. We'll post an update when it's resolved.

**Reply by wdmhouston** [reply]
thank you!

**Comment by francoishamon** [comment, marked as answer]
@wdmhouston This limitation is being fixed right now. 

If you want to try the `TableRelativePermeability` for two-phase gas-water systems without waiting, you can checkout the branch `feature/hamon/pcTable` [here](https://github.com/GEOSX/GEOSX/pull/1633). On the same branch you can also use `TableCapillaryPressure`.

The format is the following:
```
    <TableRelativePermeability                                                                                                                                                           
      name="relperm"                                                                                                                                                                     
      phaseNames="{ gas, water }"                                                                                                                                                        
      wettingNonWettingRelPermTableNames="{ waterRelativePermeabilityTable, gasRelativePermeabilityTable }"/>                                                                                                     
                                                                                       
    <TableCapillaryPressure                                                                                                                                                              
      name="cappres"                                                                                                                                                                     
      phaseNames="{ gas, water }"                                                                                                                                                        
      wettingNonWettingCapPressureTableName="waterCapillaryPressureTable"/> 
```
The documentation for the two models is here: 
https://github.com/GEOSX/GEOSX/blob/fa2b04dd78b25bc446f73a4c40dc2bc17345cb98/src/coreComponents/constitutive/docs/TableRelativePermeability.rst
https://github.com/GEOSX/GEOSX/blob/fa2b04dd78b25bc446f73a4c40dc2bc17345cb98/src/coreComponents/constitutive/docs/TableCapillaryPressure.rst

An example is attached (simulation of transition zone). 

The behavior should be similar to that of `SGFN` and `SWFN`. Let me know if you find something not working as expected.

[pc_table.tar.gz](https://github.com/GEOSX/GEOSX/files/7305393/pc_table.tar.gz)

**Reply by wdmhouston** [reply]
Thanks! Will try.
